modules/icgc-argo-workflows/checkinput/resources/usr/bin/dnaalnqc.py

```diff
@@ -0,0 +1,348 @@
+import csv
+import logging
+import sys
+from collections import Counter
+from pathlib import Path
+
+logger = logging.getLogger()
+
+
+class RowChecker:
+    """
+    Define a service that can validate and transform each given row.
+
+    Attributes:
+        modified (list): A list of dicts, where each dict corresponds to a previously
+            validated and transformed row. The order of rows is maintained.
+
+    """
+
+    VALID_FORMATS = (
+        ".bam",
+        ".cram",
+    )
+
+    def __init__(
+        self,
+        #sample_col="sample",
+        #first_col="bam_cram",
+        analysis_type_col = 'analysis_type',
+        study_id_col = 'study_id',
+        patient_col = 'patient',
+        sex_col = 'sex',
+        status_col = 'status',
+        sample_col = 'sample',
+        cram_col = 'cram',
+        crai_col = 'crai',
+        experiment_col = 'experiment',
+        genome_build_col = "genome_build",
+        analysis_json_col = 'analysis_json',
+        **kwargs,
+    ):
+        """
+        Initialize the row checker with the expected column names.
+analysis_type,study_id,patient,sex,status,sample,lane,fastq_1,fastq_2,read_group,single_end,read_group_count,analysis_json
+        Args:
+            sample_col (str): The name of the column that contains the sample name
+                (default "sample").
+            first_col (str): The name of the column that contains the first (or only)
+                FASTQ file path (default "fastq_1").
+            second_col (str): The name of the column that contains the second (if any)
+                FASTQ file path (default "fastq_2").
+            single_col (str): The name of the new column that will be inserted and
+                records whether the sample contains single- or paired-end sequencing
+                reads (default "single_end").
+
+        """
+        super().__init__(**kwargs)
+        #self._sample_col = sample_col
+        #self._first_col = first_col
+        self._analysis_type_col = analysis_type_col
+        self._study_id_col = study_id_col
+        self._patient_col = patient_col
+        self._sex_col = sex_col
+        self._status_col = status_col
+        self._sample_col = sample_col
+        self._cram_col = cram_col
+        self._crai_col = crai_col
+        self._experiment_col = experiment_col
+        self._genome_build_col = genome_build_col
+        self._analysis_json_col = analysis_json_col
+        self._seen = set()
+        self.modified = []
+
+    def validate_and_transform(self, row):
+        """
+        Perform all validations on the given row and insert the read pairing status.
+
+        Args:
+            row (dict): A mapping from column headers (keys) to elements of that row
+                (values).
+
+        """
+        #{"analysis_type","study_id","patient","sex","status","sample","cram","crai","analysis_json"}
+        self._validate_analysis_type(row) if row.get(self._analysis_type_col) else ""
```

Comment:
Added conditional test to make fields optional -
https://github.com/icgc-argo-workflows/argo-modules/pull/25#issuecomment-1832266399